staging/src/k8s.io/apiserver/pkg/storage/interfaces.go

```diff
@@ -236,6 +236,12 @@ type Interface interface {
 
 	// Count returns number of different entries under the key (generally being path prefix).
 	Count(key string) (int64, error)
+
```

Comment:
My comment is disappearing here, so I will add it here.

I still don't understand what exactly the semantic is.
Let's say etcd is RV=1000, my watch is lagging and is on RV=100 and I'm sending progress-notify.
What will happen?

Reply:
You will get RV=100. Updated the comment.

Reply:
hmm - then this whole PR doesn't really work, because I should be periodically requesting that bookmark, no?

Imagine a case from below:
- etcd is on RV=1000
- my watch is lagging and send event with RV=100 [but it has a queue of events with rv=200, rv=500, rv=999 to send]

If I know request a bookmark, from what you're saying I will get bookmark with rv=100,
then I will get rv=200, rv=500 and rv=999
but I will never get rv=1000 so I will eventually timeout.


BTW - the comment below doesn't explain the semantic you're describing above. it's still not how I would interpret it.

Reply:
> hmm - then this whole PR doesn't really work, because I should be periodically requesting that bookmark, no?

Don't agree. Requesting a bookmark was not supposed to improve handling slow watch. It's purpose is to help cases where watch cache didn't receive any events because resource it watches is rarely updated.

Lagging watch is expected to cause a timeout and requesting more bookmarks will not help. Lagging watch should go into bucket of fallback behavior. After 3 seconds of no events, should watch cache just do normal request or return timeout.

Reply:
That's not how I was thinking about this feature.

The watch slightly lagging may be happening fairly frequently - if such cases may result in watch timing out (instead of previously being handled from etcd) - that's not a tradeoff I'm ready for...

Reply:
> The watch slightly lagging may be happening fairly frequently - if such cases may result in watch timing out

Only if watch cache will not catch up in the mean time, independent from requesting bookmarks, watch events are expected to come and unblock read only when watch cache catches up to previously obtained linearizable revision. Request bookmark is only needed in cases there there are no changes on resource, thus no events.

Reply:
Discussed offline, results below.